### ISSUE TITLE
No option to specify Service Account in the new re-written ASM Module

### ISSUE THREAD

**Issue by psaini0212** [terraform-google-modules/terraform-google-kubernetes-engine#1179]
### TL;DR

Can't see any variable to specify service account while calling the ASM module.
The module picks up the Service account itself.
`From Logs - IMPERSONATE_SERVICE_ACCOUNT=asm-managed`

### Expected behavior

The Terraform should run without any issue.

### Observed behavior

As the module tries to impersonate its own Service Account, it complains about -

│ Error: local-exec provisioner error
│ 
│   with module.asm.module.cpr.module.gcloud_kubectl.null_resource.run_command[0],
│   on .terraform/modules/asm.cpr/main.tf line 231, in resource "null_resource" "run_command":
│  231:   provisioner "local-exec" {
│ 
│ Error running command 'PATH=/google-cloud-sdk/bin:$PATH
│ .terraform/modules/asm.cpr/modules/kubectl-wrapper/scripts/kubectl_wrapper.sh
│ msa europe-west2 prj-vo-digital-n-msa-gke false false
│ .terraform/modules/asm/modules/asm/scripts/create_cpr.sh asm-managed
│ regular false false
│ ': exit status 1. Output: + '[' 10 -lt 5 ']'

more logs - 

│ + gcloud container clusters get-credentials test --project
│ my-project --region europe-west2
│ Fetching cluster endpoint and auth data.
│ ERROR: (gcloud.container.clusters.get-credentials) ResponseError: code=403,
│ message=Required "container.clusters.get" permission(s) for
│ "projects/my-project/locations/europe-west2/clusters/test".
│ + cleanup
│ + rm -rf /tmp/kubectl_wrapper_11994_7500

### Terraform Configuration

```hcl
module "asm" {
  source                    = "terraform-google-modules/kubernetes-engine/google//modules/asm"
  project_id                = var.project_id
  cluster_name              = var.cluster_name
  cluster_location          = var.primary_region
  enable_cni                = false
  enable_fleet_registration = false
  enable_mesh_feature       = false
}
```

### Terraform Version

```sh
v1.0.5
```

**Comment by morgante** [comment]
Dupe of #1128. Just use provider impersonation.

**Comment by psaini0212** [comment]
Hi, sorry If I was not clear here, the problem is the module itself has picked up this service account **asm-managed** (as in the logs it says "IMPERSONATE_SERVICE_ACCOUNT=asm-managed"), now clearly this service account doesn't have the container.clusters.get" permission(s) so it fails.

Could you please suggest if I am missing something?

In my provider definition, I pass only access_token, Do I need to pass a service account as well?

    provider "google" {
      project = "{{ GCP_PROJECT_ID }}"
      access_token = "{{ DEPLOY_TOKEN }}"
      region = "{{ GCP_REGION }}"
      zone = "{{ GCP_ZONE }}"
    }

**Comment by morgante** [comment]
That shows up in the logs just because we are *checking* for that input. We don't actually impersonate [a service account](https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/blob/master/modules/asm/main.tf#L53-L65), as can be seen here.

**Comment by psaini0212** [comment]
Ok, Thank you @morgante  so does it mean the service account running the terraform will be used to call the command? i.e.

"gcloud container clusters get-credentials test --project my-project --region europe-west2"

which is returning this error -

"ERROR: (gcloud.container.clusters.get-credentials) ResponseError: code=403"
message=Required "container.clusters.get" permission(s)

In that case, the service account I am using in Terraform certainly has "Kubernetes Engine Admin" permissions which covers this permission - "container.clusters.get"

Can you guide me here please, what's missing?

It is first time I am deploying this ASM module.

**Comment by morgante** [comment]
No, in this case it will use whatever account you have configured for gcloud.

**Comment by morgante** [comment]
You should be able to get gcloud to use your token by setting the `CLOUDSDK_AUTH_ACCESS_TOKEN` environment variable: https://cloud.google.com/sdk/docs/release-notes#37100_2022-02-01

**Comment by psaini0212** [comment]
Hi @morgante I tried that, it doesn't work. I am still getting same error

One last query please - I looked at this script here -
https://github.com/terraform-google-modules/terraform-google-gcloud/blob/v3.1.0/modules/kubectl-wrapper/scripts/kubectl_wrapper.sh

It accepts multiple arguments -

CLUSTER_NAME=$1
LOCATION=$2
PROJECT_ID=$3
INTERNAL=$4
USE_EXISTING_CONTEXT=$5
ENABLE_IMPERSONATE_SERVICE_ACCOUNT=$6
IMPERSONATE_SERVICE_ACCOUNT=$7

The 6th argument is "ENABLE_IMPERSONATE_SERVICE_ACCOUNT" which is a boolean

If I check my Terraform logs it tells me that is getting passed wrong?

module.asm.module.cpr.module.gcloud_kubectl.null_resource.run_command[0] (local-exec): Executing: ["/bin/sh" "-c" "PATH=/google-cloud-sdk/bin:$PATH\n.terraform/modules/asm.cpr/modules/kubectl-wrapper/scripts/kubectl_wrapper.sh msa europe-west2 prj-vo-digital-n-msa-gke false false .terraform/modules/asm/modules/asm/scripts/create_cpr.sh asm-managed regular true false\n"]

The value that is getting passed as 6th argument is -  ".terraform/modules/asm/modules/asm/scripts/create_cpr.sh"

Doesn't that look wrong?

**Comment by bharathkkb** [comment]
@psaini0212 Could you try running `gcloud container clusters get-credentials test --project my-project --region europe-west2` to see if your gcloud is authenticated?

**Comment by psaini0212** [comment]
Hi @bharathkkb Thank you for checking it, Sir.

so this morning what I did is updated my CLoudSDK version so that it respects the "CLOUDSDK_AUTH_ACCESS_TOKEN" flag (as it was in the recent version) like it was suggested by @morgante 
Now Indeed my Terraform can run gcloud commands and I have tested the connectivity as well byt creating a null resource and it works.

resource "null_resource" "test_cluster_connectivity" {
  provisioner "local-exec" {
    command = "gcloud container clusters get-credentials test --project my-project --region europe-west2"
  }
}

Cluster Connectivity works -
null_resource.test_cluster_connectivity: Provisioning with 'local-exec'...
null_resource.test_cluster_connectivity (local-exec): Executing: ["/bin/sh" "-c" "gcloud container clusters get-credentials test --project my-project --region europe-west2"]
null_resource.test_cluster_connectivity (local-exec): Fetching cluster endpoint and auth data.
null_resource.test_cluster_connectivity (local-exec): kubeconfig entry generated for msa.
null_resource.test_cluster_connectivity: Creation complete after 2s [id=24[317](https://gitlab.com/virginmediaO2/digital/digital-msa/infrastructure/msa-gke/-/jobs/2215535878#L317)17591908669970]

My ASM deployment has even passed the last point of failure, so thank you for suggesting that variable (CLOUDSDK_AUTH_ACCESS_TOKEN)

however the logging I think still needs to be updated, it causes confusion (the logs I pasted in the last comment)

Now the ASM code is failing at this point - 

module.asm.module.cpr.module.gcloud_kubectl.null_resource.run_command[0] (local-exec): error: unable to recognize "STDIN": no matches for kind "ControlPlaneRevision" in version "mesh.cloud.google.com/v1beta1"

 Fetching cluster endpoint and auth data.
│ kubeconfig entry generated for msa.
│ + .terraform/modules/asm/modules/asm/scripts/create_cpr.sh asm-managed
│ regular true false
│ Ensuring ControlPlaneRevision exists in cluster... attempt 1
│ NAME                                          CREATED AT
│ controlplanerevisions.mesh.cloud.google.com   2022-03-17T10:37:49Z
│ customresourcedefinition.apiextensions.k8s.io/controlplanerevisions.mesh.cloud.google.com
│ condition met
│ error: unable to recognize "STDIN": no matches for kind
│ "ControlPlaneRevision" in version "mesh.cloud.google.com/v1beta1"
│ + cleanup
│ + rm -rf /tmp/kubectl_wrapper_23172_19945

Can you please suggest what's missign now?
